store/store.go

```diff
@@ -0,0 +1,415 @@
+
+	if !errors.Is(err, ErrNotFound) {
+		return nil, fmt.Errorf("getting by hash: %w", err)
+	}
+
+	// create Q1Q4 file
+	f, err = file.CreateQ1Q4File(filePath, datahash, square)
+	if err != nil {
+		return nil, fmt.Errorf("creating ODS file: %w", err)
+	}
+	return f, nil
+}
+
+func (s *Store) GetByHash(ctx context.Context, datahash share.DataHash) (eds.AccessorStreamer, error) {
+	if datahash.IsEmptyRoot() {
+		return emptyAccessor, nil
+	}
+	lock := s.stripLock.byDatahash(datahash)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	f, err := s.getByHash(datahash)
+	s.metrics.observeGet(ctx, time.Since(tNow), err != nil)
+	return f, err
+}
+
+func (s *Store) getByHash(datahash share.DataHash) (eds.AccessorStreamer, error) {
+	if datahash.IsEmptyRoot() {
+		return emptyAccessor, nil
+	}
+	path := s.basepath + blocksPath + datahash.String()
+	return s.openFile(path)
+}
+
+func (s *Store) createHeightLink(datahash share.DataHash, height uint64) error {
+	filePath := s.basepath + blocksPath + datahash.String()
+	// create hard link with height as name
+	linkPath := s.basepath + heightsPath + strconv.Itoa(int(height))
+	err := os.Link(filePath, linkPath)
+	if err != nil {
+		return fmt.Errorf("creating hard link: %w", err)
+	}
+	return nil
+}
+
+func (s *Store) GetByHeight(ctx context.Context, height uint64) (eds.AccessorStreamer, error) {
+	lock := s.stripLock.byHeight(height)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	f, err := s.getByHeight(height)
+	s.metrics.observeGet(ctx, time.Since(tNow), err != nil)
+	return f, err
+}
+
+func (s *Store) getByHeight(height uint64) (eds.AccessorStreamer, error) {
+	f, err := s.cache.Get(height)
+	if err == nil {
+		return f, nil
+	}
+	path := s.basepath + heightsPath + fmt.Sprintf("%d", height)
+	return s.openFile(path)
+}
+
+func (s *Store) HasByHash(ctx context.Context, datahash share.DataHash) (bool, error) {
+	if datahash.IsEmptyRoot() {
+		return true, nil
+	}
+	lock := s.stripLock.byDatahash(datahash)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	exist, err := s.hasByHash(datahash)
+	s.metrics.observeHas(ctx, time.Since(tNow), err != nil)
+	return exist, err
+}
+
+func (s *Store) hasByHash(datahash share.DataHash) (bool, error) {
+	if datahash.IsEmptyRoot() {
+		return true, nil
+	}
+	path := s.basepath + blocksPath + datahash.String()
+	return pathExists(path)
+}
+
+func (s *Store) HasByHeight(ctx context.Context, height uint64) (bool, error) {
+	lock := s.stripLock.byHeight(height)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	tNow := time.Now()
+	exist, err := s.hasByHeight(height)
+	s.metrics.observeHas(ctx, time.Since(tNow), err != nil)
+	return exist, err
+}
+
+func (s *Store) hasByHeight(height uint64) (bool, error) {
+	_, err := s.cache.Get(height)
+	if err == nil {
+		return true, nil
+	}
+
+	path := s.basepath + heightsPath + fmt.Sprintf("%d", height)
+	return pathExists(path)
+}
+
+func (s *Store) Remove(ctx context.Context, height uint64) error {
```

Comment:
We may have a slightly more optimized version of this method that takes both `height` and `hash.` The pruner who will be calling this has full access to the header.